src/runtime/styled.js

```diff
@@ -11,19 +13,55 @@ function omitNonHostProps(props) {
   return result;
 }
 
-export const styled = (type, displayName, styles, getStyles) => {
-  const styleMap = getStyles(styles);
+function styled(type, displayName, styles, camelName, kebabName) {
+  const componentClassName = has.call(styles, camelName)
+    ? styles[camelName]
+    : styles[kebabName];
   const omit = typeof type === 'string' ? omitNonHostProps : null;
 
-  function Styled(_props) {
-    const props = omit ? omit(_props) : { ..._props };
-    delete props.innerRef;
-    props.ref = _props.innerRef;
-    props.className = classNames(
-      props.className,
-      styleMap.map(s => (typeof s === 'string' ? s : s(_props))),
-    );
-    return React.createElement(type, props);
+  const hasModifiers = Object.keys(styles).some(
+    className => className !== camelName && className !== kebabName,
+  );
+
+  function Styled(props) {
+    const childProps = omit ? omit(props) : { ...props };
+
+    delete childProps.innerRef;
+    childProps.ref = props.innerRef;
+
+    if (hasModifiers) {
+      const modifierClassNames = [];
+
+      Object.keys(props).forEach(propName => {
+        const propValue = props[propName];
+
+        if (
+          (typeof propValue === 'boolean' || typeof propValue === 'string') &&
+          has.call(styles, propName)
+        ) {
+          if (propValue === true) {
+            modifierClassNames.push(styles[propName]);
+          } else if (propValue !== false && has.call(styles, propValue)) {
```

Comment:
maybe we should camel-case `propValue` strings (and perhaps `propName`... though anybody using non-camel-case prop names but enabling camel-casing in the css modules transform deserves what they get)